keps/sig-network/4444-service-routing-preference/README.md

```diff
@@ -0,0 +1,998 @@
+allow users to express a much desired use case from topologyKeys which is
+"prefer node-local, failover to same zone, then route anywhere" While this
+specific behavior is a non-goal for the initial implementation of the
+`routingPreference` field, the design allows for the potential introduction of
+such a preference in future refinements.
+
+## Goals
+
+* **Enhanced User Control:** Provide users with a knob to specify a preference that
+  MAY influence the traffic routing of a Kubernetes Service.
+
+* **Guidance for Implementations:** Offer standard routing preference values
+  that implementations can choose to recognize and support, promoting a degree
+  of commonality.
+
+* **Flexibility:** Allow implementations to interpret and support standard
+  heuristics in a manner that aligns with their capabilities and design.
+
+* **Extensibility:** Enable Kubernetes implementations to introduce and support
+  innovative routing strategies (for example, those based on topology, latency,
+  or other custom heuristics) that can evolve alongside users needs and
+  infrastructure capabilities.
+
+## Non-Goals
+
+* **Strict Routing Guarantees:** The `routingPreference` field will not
+  enforce deterministic routing paths. It serves as a mechanism for expressing
+  hints and preferences that implementations can consider when making routing
+  decisions.
+
+* **Mandatory and Uniform Implementation Support:** Kubernetes implementations
+  are not required to support all standard heuristics (e.g., `Zone`, `Spread`).
+  Even when standard heuristics are supported, their precise behavior and
+  interpretation might vary across implementations.
+
+* **Replacement of Traffic Policies:** The new field is complementary to
+  InternalTrafficPolicy and ExternalTrafficPolicy. It does not aim to substitute
+  their role in enforcing strict traffic locality.
+
+* **Immediate Support for All Possible Heuristics:** The initial implementation
+  focuses on a core set of heuristics. Addition of new heuristics (like
+  `Local` for Node local preference) could be explored in future
+  refinements.
+
+## Proposal
+
+Add a new field, `routingPreference`, to the Service specification. This field
+will serve as preference or hint for the underlying implementation to consider
+while making routing decisions. It does not offer strict routing guarantees.
+
+The field will support the following initial values:
+
+* `Default`: Indicates no specific routing preference. The user delegates the
+  routing decision to the implementation, allowing it to apply its best-effort
+  strategy.
+* `Spread`: Encourages an equal distribution of traffic across
+  endpoints, potentially spanning multiple zones (or regions).
+* `Zone`: Encourages routing traffic to endpoints within the same zone as
+  the client. If no endpoints are available within the zone, traffic should be
+  routed to other zones.
+
+Implementations are strongly encouraged to support the standard values. While
```

Comment:
"Strongly encoraged" != required.

Is it OK for an implementation to not support "Zone" at all?  If so, should it accept the value and ignore it, or should it fail to program endpoints at all?